README.md

```diff
@@ -283,15 +284,55 @@ enum Flags {
 
     // DataForAgentFlags is a bit mask. Values below define individual bits.
 
-    // The server asks the agent to report effective config.
+    // The server asks the agent to report effective config. This bit MUST NOT be
+    // set if the Agent indicated it cannot report effective config by setting
+    // the ReportsEffectiveConfig bit to 0 in StatusReport.capabilities field.
     ReportEffectiveConfig = 0x00000001;
 
-    // The server asks the agent to report addon statuses.
+    // The server asks the agent to report addon statuses. This bit MUST NOT be
+    // set if the Agent indicated it cannot report addon status by setting
+    // the ReportsAddonStatus bit to 0 in StatusReport.capabilities field.
     ReportAddonStatus     = 0x00000002;
 }
 ```
 
 
+#### capabilities
+
+Bitmask of flags defined by Capabilities enum. All bits that are not defined in
+Capabilities enum MUST be set to 0 by the Server. This allows extending the
+protocol and the Capabilities enum in the future such that old Servers
+automatically report that they don't support the new capability. This field MUST
+be set in the first DataForAgent sent by the Server and MAY be omitted in
+subsequent DataForAgent messages by setting it to Unspecified value.
+
+```protobuf
+enum ServerCapabilities {
+  // The capabilities field is unspecified.
+  Unspecified = 0;
+  // The Server can accept status reports. This bit MUST be set, since all Server
+  // MUST be able to accept status reports.
+  AcceptsStatus                  = 0x00000001;
+  // The Server can offer remote configuration to the Agent.
+  OffersRemoteConfig             = 0x00000002;
+  // The Server can accept EffectiveConfig in StatusReport.
+  AcceptsEffectiveConfig         = 0x00000004;
+  // The Server can offer Addons.
+  OffersAddons                   = 0x00000008;
+  // The Server can accept Addon status.
+  AcceptsAddonsStatus            = 0x00000010;
+  // The Server can offer packages to install.
+  OffersAgentPackage             = 0x00000020;
+  // The Server can accept the installation status of the package.
+  AcceptsAgentPackageStatus      = 0x00000040;
+  // The Server can offer connection settings.
+  OffersConnectionSettings       = 0x00000080;
```

Comment:
AgentCapabilities have `ReportOwnTraces` at this position and then a bunch of other capabilities. Should we keep `ServerCapabilities` in sync with `AgentCapabilities`?

Reply:
Now that I think about and look at the relationship from both ends, I see following:

* when server _offers_ something to the agent, it just advertises that certain information will be sent in the message; I don't think it changes anything (any of the "Offers" messages); 
* when server _accepts_ something from the agent, agent could pull that information and include it in the message
* when agent _reports_ something to the server, server can anticipate that certain data will be sent, which means it might e.g. wait for statuses before sending a message to the agent
* when agent _accepts_ something from the server, server can use this information to construct the message (e.g. if no addons are accepted, it might skip sending addons)

I might be missing something, but I am not sure how any of the server Offers* information can be practically used by the agent?

Reply:
> I might be missing something, but I am not sure how any of the server Offers* information can be practically used by the agent?

Exactly. You described it better than I could.

The only use that I can think of is logging on the Agent side so that when troubleshooting it is clear _why_ we are not seeing the offer from the Server. I don't know if we care about knowing this at a finer detail than just "Server can offer connection settings".

Reply:
Got it. I think it won't hurt much to keep it in the specs.

Another thing I am wondering about though is if server should provide following: `AcceptsAgentTraces`, `AcceptsAgentMetrics`, `AcceptsAgentLogs`. I believe that while own telemetry setup would be provided by server, having such information might be still useful for the agent in certain scenarios (e.g. perhaps agent would fallback to alternative config or so)

Reply:
> Another thing I am wondering about though is if server should provide following: `AcceptsAgentTraces`, `AcceptsAgentMetrics`, `AcceptsAgentLogs`.

It would be kind of misleading though, since it is not the OpAMP server that accepts traces/metrics/logs, it is probably a different server (telemetry backend), the credentials of which are offered via ConnectionSettings message. I think the ConnectionSettings message already conveys that information.

>I believe that while own telemetry setup would be provided by server, having such information might be still useful for the agent in certain scenarios (e.g. perhaps agent would fallback to alternative config or so when server does not accept some of the own telemetry signals)

The Agent can already make this decision based on the _absence_ of the particular class of ConnectionSettings offer from the Server (e.g. if ConnectionSettingsOffers.own_traces field is unset by the Server). I think the additional capabilities bit is not necessary for that.